compiler/rustc_lexer/src/lib.rs

```diff
@@ -642,40 +644,79 @@ impl Cursor<'_> {
                 '.' | 'e' | 'E' => {}
 
                 // Just a 0.
-                _ => return Int { base, empty_int: false },
+                _ => {
+                    return (Int { base, empty_int: false }, None);
+                }
             }
         } else {
             // No base prefix, parse number in the usual way.
             self.eat_decimal_digits();
         };
 
-        match self.first() {
+        match (self.first(), self.second()) {
             // Don't be greedy if this is actually an
             // integer literal followed by field/method access or a range pattern
             // (`0..2` and `12.foo()`)
-            '.' if self.second() != '.' && !is_id_start(self.second()) => {
+            ('.', second) if second != '.' && !is_id_start(second) => {
                 // might have stuff after the ., and if it does, it needs to start
                 // with a number
                 self.bump();
                 let mut empty_exponent = false;
+                let mut suffix_start = None;
                 if self.first().is_ascii_digit() {
                     self.eat_decimal_digits();
-                    match self.first() {
-                        'e' | 'E' => {
+                    // This will be the start of the suffix if there is no exponent
+                    suffix_start = Some(self.pos_within_token());
+                    match (self.first(), self.second()) {
+                        ('e' | 'E', '_') => {
+                            // check if series of `_` is ended by a digit. If yes
+                            // include it in the number as exponent. If no include
+                            // it in suffix.
+                            self.bump();
+                            while matches!(self.first(), '_') {
+                                self.bump();
+                            }
+                            // If we find a digit, then the exponential was valid
+                            // so the suffix will start at the cursor as usual.
+                            if self.first().is_ascii_digit() {
```

Comment:
Simpler to do an early return on `!self.first().is_ascii_digit()`, that's the only case when the suffix start is not `None`, and `suffix_start` mutation is not necessary.

Reply:
I think I've addressed - please LMK if I misunderstood.